src/Nethermind/Nethermind.Synchronization/Peers/PeerInfo.cs

```diff
@@ -19,14 +20,33 @@ namespace Nethermind.Synchronization.Peers
 {
     public class PeerInfo
     {
-        public PeerInfo(ISyncPeer syncPeer)
+        public static readonly Dictionary<AllocationContexts, int> DefaultAllowances = new Dictionary<AllocationContexts, int>()
+        {
+            {AllocationContexts.Headers, 1},
+            {AllocationContexts.Bodies, 1},
+            {AllocationContexts.Receipts, 1},
+            {AllocationContexts.State, 1},
+            {AllocationContexts.Snap, 1},
+        };
+
+        private readonly Dictionary<AllocationContexts, int> _allocationAllowances;
+
+        public PeerInfo(ISyncPeer syncPeer, Dictionary<AllocationContexts, int>? allocationAllowances = null)
         {
             SyncPeer = syncPeer;
+
+            allocationAllowances ??= DefaultAllowances;
+            AllocationSlots = new ConcurrentDictionary<AllocationContexts, int>(allocationAllowances);
+            _allocationAllowances = allocationAllowances;
         }
 
         public NodeClientType PeerClientType => SyncPeer?.ClientType ?? NodeClientType.Unknown;
 
-        public AllocationContexts AllocatedContexts { get; private set; }
+        public ConcurrentDictionary<AllocationContexts, int> AllocationSlots { get; init; }
+
+        public Dictionary<AllocationContexts, int> AvailableAllocationSlots => AllocationSlots
+            .Select(kv => (kv.Key, _allocationAllowances[kv.Key] - kv.Value))
+            .ToDictionary();
```

Comment:
simple array would be better? Or in-line array even.
You could ref and Interlock.Increment it.

Reply:
Swtiched to enumerable